### PR TITLE
Value traits and sort filter

### PR DESCRIPTION
**Pull request by Albibek**
Hi.
- Implemented PartialEq correctly and completely, as it was asked in TODO 
- Added Eq and Ord for Value
- Made `sort` filter with test

### CODE REVIEW

File: src/filters.rs [hunk 1]
```diff
@@ -299,6 +299,23 @@ pub fn join(input: &Value, args: &[Value]) -> FilterResult {
     }
 }
 
+pub fn sort(input: &Value, args: &[Value]) -> FilterResult {
+    if args.len() > 0 {
+        return Err(FilterError::InvalidArgumentCount(format!("expected no arguments, {} given",
```
Comment (johannhof) [inline src/filters.rs]: nit: you can leave out the `FilterError::` here

File: src/filters.rs [hunk 2]
```diff
@@ -299,6 +299,23 @@ pub fn join(input: &Value, args: &[Value]) -> FilterResult {
...
+                                                             args.len())));
+    }
+    match input {
+        &Value::Array(ref array) => {
+            let mut sorted = array.clone();
+            sorted.sort();
+            Ok(Value::Array(sorted))
+        }
+        _ => Err(FilterError::InvalidType("Array argument expected".to_owned())),
```
Comment (johannhof) [inline src/filters.rs]: nit: you can leave out the `FilterError::` here

File: tests/filters.rs [hunk 3]
```diff
@@ -253,13 +252,13 @@ pub fn split_no_comma() {
 
 #[test]
 // Split on 1 string and re-join on another
-pub fn split_then_join() {
```
Comment (johannhof) [inline tests/filters.rs]: Please create a new test instead of overwriting this one :)

File: src/value.rs [hunk 4]
```diff
@@ -61,48 +61,24 @@ impl PartialEq<Value> for Value {
     }
 }
 
-// TODO implement for object and array
-// TODO clean this up
+impl Eq for Value {}
```
Comment (johannhof) [inline src/value.rs]: I suppose we can't use `derive` instead?

File: src/value.rs [hunk 5]
```diff
@@ -61,48 +61,24 @@ impl PartialEq<Value> for Value {
...
-            _ => false,
-        }
-    }
-    fn ge(&self, other: &Value) -> bool {
-        match (self, other) {
-            (&Value::Num(x), &Value::Num(y)) => x.ge(&y),
-            (&Value::Str(ref x), &Value::Str(ref y)) => x.ge(y),
-            (&Value::Bool(x), &Value::Bool(y)) => x.ge(&y),
-            _ => false,
-        }
```
Comment (johannhof) [inline src/value.rs]: Nice!
This reminds me: Can you please add a couple of assertions to the `number_equality` test that verify that `<`, `>`, `<=` and `>=` comparisons work correctly? Thanks!

File: src/value.rs [hunk 6]
```diff
@@ -61,48 +61,24 @@ impl PartialEq<Value> for Value {
...
+}
+
+impl Ord for Value {
+    fn cmp(&self, other: &Value) -> Ordering {
+        self.partial_cmp(other).unwrap_or(Ordering::Equal)
```
Comment (johannhof) [inline src/value.rs]: Do we really want to fall back to equal here?

**Comment by Albibek** [comment]
About ordering. That's my bad of thinking about this only in terms of sorting. I think it would me more correct to remove the Ord impl and move the same like that to sort filter using `sort_by` instead of `sort`. It should give the intented level of comparison while sorting the array where all types are same while not touching the global inconsistence as it was with my approach.
We could even support partial equality of floats adding some loginc and returning error from sort filter, but I'm not sure we really need such level of correctness.
What do you think?

**Comment by johannhof** [comment]
> I think it would me more correct to remove the Ord impl and move the same like that to sort filter using sort_by instead of sort. It should give the intented level of comparison while sorting the array where all types are same while not touching the global inconsistence as it was with my approach.

That sounds great to me.

> We could even support partial equality of floats adding some loginc and returning error from sort filter, but I'm not sure we really need such level of correctness.

Not sure if I understand you correctly, but it sounds unnecessary :)

Thanks!

**Comment by Albibek** [comment]
I meant that floats, for example, don't impl Ord, because NaNs are not orderable. Anyways that is rare case, having no meaning in templates. See last commit for the fix.

**Comment by johannhof** [comment]
I see, thanks for the explanation. This looks good to me now. :)

**Pull request merged**